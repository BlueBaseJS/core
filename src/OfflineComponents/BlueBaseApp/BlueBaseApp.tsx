import { BlueBase, BlueBaseProgress, BootOptions } from '../../BlueBase';
import {
	BlueBaseAppError,
	BlueBaseAppErrorProps,
	BlueBaseAppLoading,
	BlueBaseAppLoadingProps,
} from '..';
import React, { useEffect, useState } from 'react';

import { BlueBaseProvider } from '../../Context';
import { IntlProvider } from '../../intl';
import { ThemeProvider } from '../../themes';
import { getComponent } from '../../getComponent';

const BlueBaseContent = getComponent('BlueBaseContent');

export interface BlueBaseAppProps extends Partial<BootOptions> {
	/** BlueBase Context */
	BB?: BlueBase;

	/** Component rendered when an error occurs */
	ErrorComponent: React.ComponentType<BlueBaseAppErrorProps>;

	/** Component rendered when an error occurs */
	LoadingComponent: React.ComponentType<BlueBaseAppLoadingProps>;

	/**
	 * If this prop is provided, this tree will be rendered instead of BlueBase's own view.
	 */
	children?: React.ReactNode;

	/**
	 * Used to locate this view in end-to-end tests.
	 */
	testID?: string;
}

/**
 * # ðŸš€ BlueBaseApp
 *
 * The main BlueBase app. This is the top level component in BlueBase. Takes care
 * of initialisation, and renders either children, or app with routing.
 *
 * ## Usage
 * ```jsx
 * <BlueBaseApp BB={BB} />
 * ```
 */
export const BlueBaseApp = (props: BlueBaseAppProps) => {
	const {
		ErrorComponent = BlueBaseAppError,
		LoadingComponent = BlueBaseAppLoading,
		children,
	} = props;

	const [bootCount, setBootCount] = useState(0);
	const [booting, setBooting] = useState(true);
	const [bootTrigger, setBootTrigger] = useState(true); // Setting to true to start boot
	const [progress, setProgress] = useState<BlueBaseProgress>({});

	const [BB] = useState(props.BB || new BlueBase());

	BB.reboot = async () => {
		setBootTrigger(true);
	};

	useEffect(() => {
		if (bootTrigger) {
			setBootTrigger(false);
		} else {
			return;
		}

		(async () => {
			await BB.boot({
				...props,
				onProgress: (p: BlueBaseProgress) => {
					setProgress(p);
					if (p.error) {
						setBooting(false);
					}
				},
			});
			setBootCount(bootCount + 1);
			setBooting(false);
		})();
	}, [bootTrigger]);

	if (booting) {
		return <LoadingComponent BB={BB} progress={progress} bootCount={bootCount} />;
	}

	if (progress.error) {
		return <ErrorComponent BB={BB} progress={progress} bootCount={bootCount} />;
	}

	return (
		<BlueBaseProvider key={`boot-${bootCount}`} value={BB}>
			<ThemeProvider>
				<IntlProvider>
					<BlueBaseContent>{children}</BlueBaseContent>
				</IntlProvider>
			</ThemeProvider>
		</BlueBaseProvider>
	);
};

const defaultProps: Partial<BlueBaseAppProps> = {
	ErrorComponent: BlueBaseAppError,
	LoadingComponent: BlueBaseAppLoading,
};

BlueBaseApp.defaultProps = defaultProps;
