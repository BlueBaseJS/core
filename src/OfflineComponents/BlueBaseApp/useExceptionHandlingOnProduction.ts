import { BlueBase } from '../../BlueBase';
import { isProduction } from '../../utils';

// Number of errors to save
const CACHE_SIZE = 5;

// If a successive error take longer than this time, we don't consider it as recursive
const RECURSIVE_PERIOD = 300;

let errors: Array<{
	error: any;
	time: number;
}> = [];

/**
 * If the app is in production mode, automatically reboots BB on exception.
 * Shows an error message if the app is stuck in infinite crash.
 * @param BB
 */
export function useExceptionHandlingOnProduction(BB: BlueBase) {
	let development = BB.Configs.getValue('development');

	if (development === undefined) {
		development = !isProduction();
	}

	/**
	 * Adds an error to the errors array, with the timestamp
	 * @param e
	 */
	function addError(e: any) {
		const arr = [
			...errors,
			{
				error: e,
				time: new Date().getTime(),
			},
		];

		errors = arr.length > CACHE_SIZE ? arr.slice(1) : arr;
	}

	/**
	 * Gets periods of time in successive errors
	 */
	function isInfinite() {
		if (errors.length < CACHE_SIZE) {
			return false;
		}

		for (let i = 1; i < errors.length; i++) {
			if (errors[i].time - errors[i - 1].time > RECURSIVE_PERIOD) {
				return false;
			}
		}

		return true;
	}

	function onError(e: any) {
		addError(e);

		if (!development && !isInfinite()) {
			BB.reboot();
		}
	}

	return { onError };
}
