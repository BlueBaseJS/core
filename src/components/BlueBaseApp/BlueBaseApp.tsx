import { BlueBase, BootOptions } from '../../BlueBase';
import React from 'react';
import { Text } from 'react-native';

export interface BlueBaseAppProps extends Partial<BootOptions> {
	BB?: BlueBase,
	children?: React.ReactNode,
}

interface BlueBaseAppState {

	booted: boolean;
	BB: BlueBase

	Component: React.ComponentType<any>
}

/**
 * ðŸš€ BlueBaseApp
 *
 * This is the Main App used to render at the top level.
 *
 * TODO: Add better docs
 */
// NOTE FOR DEVELOPERS:
// Don't use BlueBase context or any data saved in context (e.g. components)
// here. This is because the context may not be initialized yet.
export class BlueBaseApp extends React.Component<BlueBaseAppProps, BlueBaseAppState> {

	constructor(props: BlueBaseAppProps) {
		super(props);

		this.state = {
			BB: props.BB || new BlueBase(),
			Component: () => <Text>Loading</Text>,
			booted: false,
		};
	}

	async componentDidMount() {

		const BB = this.state.BB;

		try {
			const Component = await BB.boot(this.props);
			this.setState({
				Component,
				booted: BB.booted,
			});
		} catch (error) {
			// tslint:disable-next-line:no-console
			console.error(error);
		}
	}

	render() {

		const Component = this.state.Component;

		return <Component children={this.props.children} />;
	}
}