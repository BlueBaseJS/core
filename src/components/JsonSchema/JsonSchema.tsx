import { JsonComponentNode, JsonSchemaParser } from '../../lib/json-schema-parser';
import { BlueBase } from '../../BlueBase';
import { BlueBaseContext } from '../../Context';
import { BlueBaseFilter } from '../../index';
import { MaybeArray } from '../../utils';
import React from 'react';

export interface JsonSchemaProps {

	/** JSON Schema. */
	schema: MaybeArray<JsonComponentNode>;

	/** Event name to filter this schema. If this is not provided, the schema is not filtered. */
	filter?: string;

	/** Arguments for the filter. */
	args?: { [key: string]: any };

  /**
   * Used to locate this view in end-to-end tests.
   */
	testID?: string,
}

const getComponent = (BB: BlueBase) => {
	return ({ component }: JsonComponentNode) => {

		return BB.Components.has(String(component)) ? BB.Components.resolve(String(component)) : null;
	};
};

/**
 * # üç± JsonSchema
 *
 * Renders a Component based on JSON schema. This allows developers to create dynamic
 * layouts in their apps, and even save the schema to databases.
 *
 * Moreover, it also makes that schema filter-able. So that any plugin can modify that schema
 * on runtime.
 *
 * ## Usage:
 * ```jsx
 * <JsonSchema
 * 	filter="content-filter"
 * 	args={{ style: { color: 'blue' } }}
 *  schema={{
 * 	 component: 'Text',
 * 	 props: {
 * 		 style: {
 * 			 color: 'red'
 * 		 }
 * 	 },
 * 	 text: 'This is the page content.',
 *  }
 * } />
 * ```
 */
export class JsonSchema extends React.PureComponent<JsonSchemaProps> {

	static contextType = BlueBaseContext;

	render() {

		const BB: BlueBase = this.context;

		const { filter, schema, args } = this.props;
		const parser = new JsonSchemaParser(getComponent(BB));

		// There's no filter, we don't need to do complex async handling
		if (!filter) {
			return parser.parseSchema(schema);
		}

		const children = (loadedSchema: MaybeArray<JsonComponentNode>) => parser.parseSchema(loadedSchema);

		return (
			<BlueBaseFilter
				filter={filter}
				value={schema}
				args={args}
				children={children}
			/>
		);
	}
}
