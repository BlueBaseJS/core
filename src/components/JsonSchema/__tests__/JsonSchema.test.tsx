import { BlueBaseApp } from '../../BlueBaseApp';
import { JsonSchema } from '../JsonSchema';
import React from 'react';
import TestRenderer from 'react-test-renderer';

// beforeEach(() => {
// 	jest.resetModules();
// });

describe('JsonSchema', () => {

	test(`should render a text component with red color`, (done) => {
		const component = TestRenderer.create(
			<BlueBaseApp>
				<JsonSchema schema={{
					component: 'Text',
					props: {
						style: {
							color: 'red'
						}
					},
					text: 'This component is generated through JsonSchema Component',
				}} />
			</BlueBaseApp>
		);
		let tree = component.toJSON();
		expect((tree as any).children.join()).toBe('Loading');
		expect(tree).toMatchSnapshot();


		setTimeout(() => {
			tree = component.toJSON();
			expect((tree as any).children[0].props.style[1]).toMatchObject({
				color: 'red'
			});
			expect((tree as any).children[0].children.join()).toBe('This component is generated through JsonSchema Component');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should render a text component with red color that is processed by hooks`, (done) => {

		const hook = (schema: any, args: { style: any }) => ({
			...schema,
			props: {
				...schema.props,
				style: {
					...schema.props.style,
					...args.style,
				}
			},
			text: 'Content changed by hook!',
		});

		const component = TestRenderer.create(
			<BlueBaseApp hooks={{ 'content-hook': hook }}>
				<JsonSchema
					hook="content-hook"
					args={{ style: { color: 'blue' } }}
					schema={{
						component: 'Text',
						props: {
							style: {
								color: 'red'
							}
						},
						text: 'This component is generated through JsonSchema Component',
					}
				} />
			</BlueBaseApp>
		);

		setTimeout(() => {
			const tree = component.toJSON();
			expect((tree as any).children[0].props.style[1]).toMatchObject({
				color: 'blue'
			});
			expect((tree as any).children[0].children.join()).toBe('Content changed by hook!');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should render a null for unknown component`, (done) => {

		const component = TestRenderer.create(
			<BlueBaseApp>
				<JsonSchema schema={{
					component: 'Foo',
					text: 'This component is generated through JsonSchema Component',
				}} />
			</BlueBaseApp>
		);


		setTimeout(() => {
			const tree = component.toJSON();
			expect(tree).toMatchSnapshot();
			expect((tree as any).children.join())
				.toBe('Could not parse React JSON Schema. Reason: Could not resolve component.');
			done();
		});
	});

});
