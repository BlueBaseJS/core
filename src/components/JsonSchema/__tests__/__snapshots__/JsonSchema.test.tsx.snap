// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonSchema Snapshot ComponentState component 1`] = `
<Text>
  Loading
</Text>
`;

exports[`JsonSchema Snapshot ComponentState component after complete rendering 1`] = `
<View
  styles={Object {}}
>
  <Text
    style={
      Array [
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontFamily": "Roboto",
          "fontSize": 16,
          "fontWeight": "400",
        },
        Object {
          "color": "red",
        },
      ]
    }
  >
    This componenet is generated through JsonSchema Component
  </Text>
</View>
`;
